### ISSUE TITLE
Generate migration files from models and execute SQL.

### ISSUE THREAD

**Issue by eji** [sequelize/sequelize#120]
I'd like to generate migration files from models.

Now, i create migration files by hand. So, gap may occur between models and migrate files.
In addition,  a function to execute SQL is necessary for data migration, because  models may not compatible with migration files.

**Comment by milovanderlinden** [comment]
#390 is a duplicate, suggesting to create a function like rake db:schema:dump

**Comment by saikat** [comment]
For what it's worth, Django implements the ability to auto-generate migration files from model diffs.  It may be worth looking into their codebase to see how they are implementing it, it's done fairly well.

**Comment by mickhansen** [comment]
@saikat It's a matter of actually doing the work, coming up with the design is pretty straight forward

**Comment by JamieCressey** [comment]
+1 generating migrations from model diffs would be a great feature

**Comment by ArnaudValensi** [comment]
+1

**Comment by AlexHayton** [comment]
@mickhansen do you know if anyone is actively working on this? Do you have an idea for how the design should look or are you still fielding proposals?

**Comment by mickhansen** [comment]
@AlexHayton No one is to my knowledge. Implementing it comes down to a few steps. First we want to make sure that describeTable can handle almost all cases, and that it maps back 1:1 to a model it was defined from. Then we need some sort of intermediate format for diffs, and then we need to be able to generate SQL or migrations from that format.

**Comment by sukrit007** [comment]
+1 for this feature.  I used to use [liquibase](http://www.liquibase.org/documentation/generating_changelogs.html) in java world  which had an ability to generate change log based on current tables.  Really looking forward for this feature.

**Comment by felixfbecker** [comment]
Just looking at the database would not work though if you renamed a column. There is no way to find out that a column got renamed (and not removed and and another one added).

A different approach that sequelize-cli could use is doing a diff with git to see what has changed. It can use git to output the model file at the previous commit or tag and see what has changed.

**Comment by Pomax** [comment]
it can just do the exact same thing that Django's migrate (formerly South) does: it knows the current schema (from the db), and it knows the the new schema. When generating the migration, if it's unsure, it simply asks. "Unable to detect operation. Did you rename <colname> to <colname>? [y/n]". As an explicit and infrequent operation, asking the user to make the decision is perfectly fine.

**Comment by felixfbecker** [comment]
@Pomax asking the user does not work in Continuous Deployment ;)

**Comment by Pomax** [comment]
it certainly does, because you generate the migration outside of the deployment. It's part of the commit, run on the committer's machine, alongside the PR for the schema change. You make the migration as part of your schema change work, at which point you can get the input necessary to build the correct migration script. Then you commit that and CI works perfectly fine because there's no ambiguity anymore - it just runs the migration.

**Comment by felixfbecker** [comment]
You are right, I thought you meant asking the user when the migration is run.

**Comment by zuck** [comment]
+1 for Django-style migration system.

**Comment by 13pass** [comment]
+1

**Comment by felixfbecker** [comment]
Please don't +1 issues. It clutters the thread without adding value to the discussion and spams maintainers with notifications. Use GitHub reactions to upvote features.
![image](https://cloud.githubusercontent.com/assets/10532611/18344735/bb60c2f6-75b8-11e6-88a9-822fae298110.png)

**Comment by normano** [comment]
@durango @mickhansen 

I am surprised no one has attempted their hand at this or at least has come up with something past generate migrations only from current model state.

I have a rough script (bugs ofc) to basically do this but only for add/remove tables and columns (I'll have changeColumn when I require it or if someone else can contribute). I am using Sequelize 4, so not sure if my script will work on version 3.

The idea is to create a temporary database, run previous migrations on it and then run a diff using the temporary database vs the models and spit out migration files that will get from the previous migration to the state that we want.

https://gist.github.com/normano/d41df9d5491d5aca89345e3623b76f6d

You guys are more than welcome to improve the code to bring it from rough code to professional tested code.

**Comment by jperl** [comment]
The workaround I am using for now is

```
export default async function createTables() {
  const sql = [];

  await connection.sync({
    force: true,
    logging: (message) => {
      let line = message.replace('Executing (default): ', '').trim();
      if (line.substr(-1) !== ';') {
        line += ';';
      }

      sqlLines.push(line);
    },
  });

  const sql = sqlLines.join('\n');
  fs.writeFile('./migrations/initial-tables.sql', sql);
```

```
const fs = require('fs');

module.exports = {
  up(migration) {
    return new Promise((resolve, reject) => {
      fs.readFile('migrations/initial-tables.sql', 'utf8', (error, sql) => {
        if (error) {
          reject(error);
          return;
        }

        migration.sequelize
          .query(sql, { raw: true })
          .then(resolve);
      });
    });
  },

  down(queryInterface, sequelize) {
    return queryInterface.dropAllTables();
  },
};
```

**Comment by brud** [comment]
I found this project https://github.com/flexxnn/sequelize-auto-migrations - it look likes perspective

**Comment by murbanowicz** [comment]
Any updates on this?

**Comment by Luomusha** [comment]
About 6 years passed.This issue still alive,and no progress.

**Comment by botaniQQQ** [comment]
7 years ...

**Comment by mickaelw** [comment]
up

**Comment by papb** [comment]
@murbanowicz @Luomusha @botaniQQQ @mickaelw Please check [sequelize-auto-migrations](https://github.com/flexxnn/sequelize-auto-migrations) as brud said.

**Comment by mysuf** [comment]
@papb Unfortunately, that project is not maintained for some time and looks dead..

**Comment by papb** [comment]
@mysuf Hmmm... That's unfortunate... So perhaps there is no up-to-date workaround then.

**Comment by agodin3z** [comment]
Hello, looking for the solution to this same problem I ran into a Gist, I modified it to work for me ... I hope it helps someone too ^^

https://gist.github.com/agodin3z/4b87a62863bf676cb6a8420fec0cd4b8

**Comment by neekey** [comment]
switching from SQLAlchemy using its migration tool https://flask-migrate.readthedocs.io/en/latest/ (Python), I feel like a migration auto-detection feature is not just nice to have, it's like a must.

**Comment by easherma** [comment]
+ 1, coming from Python, SQLAlchemy or Django both handle this quite nicely.

**Comment by snake-py** [comment]
I am actually quite surprised that this is not in there. How can we contribute to the package?

**Comment by schieck0** [comment]
9 years after...

**Comment by Axel-AMGHAR** [comment]
Comming from symfony where it's handled (and well handled in laravel as well), I literaly don't understand why you have to write 2 times in models in migrations. It's seems such a basic thing to implement. And this issue is been going for so long. Im far for being an expert but this issue is running for so long, it's quite frustationg to spend weeks searching for a solution and not finding anything good enougth. I wasted kind of a lot of time.

**Comment by mrfy** [comment]
Such functionality would be very useful :)

**Comment by ricardojr** [comment]
Yes, the above solution found by @agodin3z is interesting. It requires some tweaking to meet each project's needs but it is a good starting point. 

@sidneyarcidiacono I read some suggestions in this thread about syncing the schema automatically, etc. Even though this might be supported by other platforms, I think the ROI is not worth it. However, the script referenced above seems very easy to maintain and valuable, thus has a much higher ROI. Even if it is simply adding it to the documentation under advanced concepts as a reference snippet, to serve as an example of how such functionality could be implemented to each project's need.

**Comment by faezehnm** [comment]
9 years after :))))

**Comment by jking6884** [comment]
Is there no intention to add this functionality which seems almost expected in similar solutions by today's standards?

**Comment by WikiRik** [comment]
This is what https://github.com/sequelize/sequelize-auto adds right?

**Comment by jking6884** [comment]
> [<img alt="@stale" width="20" height="20" src="https://avatars.githubusercontent.com/in/1724?s=60&amp;v=4">](https://github.com/marketplace/stale) [stale](https://github.com/marketplace/stale) bot added the [ stale ](/sequelize/sequelize/labels/stale) label [on Jun 29, 2017](#event-1143940836)
> [<img alt="@jperl" width="40" height="40" src="https://avatars.githubusercontent.com/u/1136652?s=88&amp;u=a92d3fb7dae34e5878dcb9a41196e0b18feac6e8&amp;v=4">](/jperl)

@WikiRik I believe sequelize-auto is trying to do the exact opposite.

The functionality this thread is talking about is similar to django framework and so many other frameworks which are mentioned in this thread.

The way it works in Django, for instance, you define (or make changes to) your model file and then you run a command like:
`python manage.py makemigrations`

This command will read all of the models and do a diff to determine which models have new changes since last this method was called.  When it detects changes, it will create a migration file.  Then after that you run:

`python manage.py migrate`

This will then execute the migrations and modify the database

**Comment by VaZark** [comment]
It'd be nice if we didn't have to write the exact same Model and migration logic twice over as it increases the chances of dev error. Can the migration script be ignored and replaced with plain SQL like diesel in rust?

**Comment by ephys** [comment]
@VaZark Umzug is capable of doing that but I don't think anything of the sort has been added to sequelize-cli. You can open a feature request on its repository if you want https://github.com/sequelize/cli

**Comment by amonacom** [comment]
10 years! Happy anniversary issue! The need for this feature is there, check out  this libraries and their myriads of forks:

https://github.com/matmar10/sequelize-lazy-migrations
https://github.com/Scimonster/sequelize-auto-migrations
https://github.com/flexxnn/sequelize-auto-migrations

It would be probably a good idea to incorporate this functionality in the core and finally close this issue?

**Comment by towfiqi** [comment]
For Typescript use https://www.npmjs.com/package/sequelize-typescript-migration-lts

**Comment by euqen** [comment]
As an alternative you could use atlas to generate schema migrations sql: https://atlasgo.io/guides/orms/sequelize

This will allow you to generate SQL versioned migrations. On top of this, assuming you use umzug as a migration tool which is native for sequelize users, it could run different migrations: written in js,ts and pure SQL. More info here: 

https://github.com/sequelize/umzug/tree/main?tab=readme-ov-file#additional-migration-configuration-options

**Comment by adityasingh2640** [comment]
It would be nice if we can have auto generated migration files. I recently worked on Nest.js and Typeorm project, there we can generate migration files just by simple cli commands. Looking for some same soultion here also.

**Comment by Canoir** [comment]
1+

**Comment by oxydeon** [comment]
On a NestJS project, I managed to use typescript models declarations in migation of create table like this:

- Declaration of model
```
import { Column, DataType, Model, Table } from 'sequelize-typescript';

export class EntityExample extends Model {
  @Column({
    primaryKey: true,
    type: DataType.UUID,
    defaultValue: DataType.UUIDV4,
  })
  id: string;

  @Column({
    allowNull: false,
    unique: true,
  })
  name: string;
}
```

- Extraction of model
```
// Create a new instance of sequelize to be able to instanciate models and retrieve table information
const sequelize = new Sequelize({ dialect: '<dialect>') });

export function getTableInfo(table): {
  schema?: string;
  tableName: string;
  attributes: { [key: string]: ModelAttributeColumnOptions };
} {
  sequelize.addModels([table]);
  const { schema, tableName } = table.getTableName() as TableNameWithSchema;

  return {
    schema,
    tableName,
    attributes: table.getAttributes(),
  };
}
```

- Migration
```
const { schema, tableName, attributes } = getTableInfo(table);

module.exports = {
  async up({ context }: { context: QueryInterface }): Promise<void> {
    await context.createTable(
      { schema, tableName },
      attributes,
    );
  },

  async down({ context }: { context: QueryInterface }): Promise<void> {
    await context.dropTable({ schema, tableName });
  },
};
```
